### PR TITLE
Add params_with_assocs function

### PR DESCRIPTION
**Pull request by jsteiner**
Addresses: https://github.com/thoughtbot/ex_machina/issues/105

### CODE REVIEW

File: lib/ex_machina/ecto.ex [hunk 1]
```diff
@@ -66,6 +70,33 @@ defmodule ExMachina.Ecto do
     |> drop_ecto_fields
   end
 
+  @doc """
+  Same as `params_for/2`, but it inserts all associations and sets the
+  foreign_keys.
```
Comment (paulcsmith) [inline lib/ex_machina/ecto.ex]: I think that it should only insert the `belongs_to` associations and not `has_many`. What do you think?

`Same as `params_for/2`, but it inserts all belongs_to associations and sets the foreign keys`

File: lib/ex_machina/ecto.ex [hunk 2]
```diff
@@ -66,6 +70,33 @@ defmodule ExMachina.Ecto do
...
+      def article_factory do
+        %MyApp.Article{title: "An Awesome Article", author: build(:user)}
+      end
+
+      # Returns %{title: "An Awesome Article", author_id: 12}
+      params_with_assocs(:article)
+  """
+  def params_with_assocs(module, factory_name, attrs \\ %{}) do
+    module.build(factory_name, attrs)
+    |> insert_ecto_fields(module)
```
Comment (paulcsmith) [inline lib/ex_machina/ecto.ex]: Maybe `insert_belongs_to_associations`?

File: lib/ex_machina/ecto.ex [hunk 3]
```diff
@@ -66,6 +70,34 @@ defmodule ExMachina.Ecto do
...
+    |> drop_ecto_fields
+  end
+
+  defp insert_ecto_fields(record = %{__struct__: struct, __meta__: %{__struct__: Ecto.Schema.Metadata}}, module) do
+    Enum.reduce(struct.__schema__(:associations), record, fn(association, record) ->
+      foreign_key = struct.__schema__(:association, association).owner_key
+      relation = Map.get(record, association) |> module.insert
```
Comment (jsteiner) [inline lib/ex_machina/ecto.ex]: This is going to break for `has_many`s. Fixing.

File: test/ex_machina/ecto_test.exs [hunk 4]
```diff
@@ -67,4 +82,17 @@ defmodule ExMachina.EctoTest do
...
+    params = Factory.params_with_assocs(:article)
+
+    assert Map.keys(params) == [:author_id, :title]
+
+    %{
+      title: "My Awesome Article",
+      author_id: author_id,
+    } = params
+
+    assert author_id != nil
```
Comment (paulcsmith) [inline test/ex_machina/ecto_test.exs]: I think this might be easier to visualize if we asserted against an exact map

```elixir
assert params == %{
  title: "My Awesome Article",
  author_id: TestRepo.one!(User).id
}
```

Comment (jsteiner) [inline test/ex_machina/ecto_test.exs]: Cool. I dig it.

File: test/ex_machina/ecto_test.exs [hunk 5]
```diff
@@ -10,13 +10,28 @@ defmodule ExMachina.EctoTest do
     end
   end
 
+  defmodule Article do
+    use Ecto.Schema
+    schema "articles" do
+      field :title, :string
+      belongs_to :author, User
```
Comment (paulcsmith) [inline test/ex_machina/ecto_test.exs]: I think it would also be good to add another `belongs_to` that isn't defined in the factory to make sure there are no issues when the association is not set

File: lib/ex_machina/ecto.ex [hunk 6]
```diff
@@ -66,6 +70,34 @@ defmodule ExMachina.Ecto do
...
   end
 
+  @doc """
+  Same as `params_for/2`, but it inserts all associations and sets the
+  foreign_keys.
+
+  ## Example
+
+      def article_factory do
+        %MyApp.Article{title: "An Awesome Article", author: build(:user)}
```
Comment (paulcsmith) [inline lib/ex_machina/ecto.ex]: Maybe `build(:author)` so the terminology remains the same throughout the example?

File: lib/ex_machina/ecto.ex [hunk 7]
```diff
@@ -66,6 +70,34 @@ defmodule ExMachina.Ecto do
...
+      end
+
+      # Inserts an author and returns
+      # %{title: "An Awesome Article", author_id: 12}
+      params_with_assocs(:article)
```
Comment (paulcsmith) [inline lib/ex_machina/ecto.ex]: I like this example a lot

File: lib/ex_machina/ecto.ex [hunk 8]
```diff
@@ -66,6 +70,34 @@ defmodule ExMachina.Ecto do
...
+  """
```
Comment (paulcsmith) [inline lib/ex_machina/ecto.ex]: It might be worth adding a note to `params_for/2` that belongs_to associations are not included and they should look at `params_with_assocs/2` if they need it. Otherwise I'm not sure people will realize it's here or why it matters

**Comment by paulcsmith** [comment]
This is awesome! I was actually going to ask if you had time to tackle this next :D

File: lib/ex_machina/ecto.ex [hunk 9]
```diff
@@ -66,6 +73,50 @@ defmodule ExMachina.Ecto do
     |> drop_ecto_fields
   end
 
+  @doc """
+  Same as `params_for/2`, but it inserts all belongs_to associations and sets the
```
Comment (paulcsmith) [inline lib/ex_machina/ecto.ex]: Maybe this could be shortened to `Same as `params_for/2`, but inserts all belongs_to associations and sets the foreign keys`? If you don't like that I'm fine with it as-is as well

File: lib/ex_machina/ecto.ex [hunk 10]
```diff
@@ -66,6 +73,50 @@ defmodule ExMachina.Ecto do
...
+  foreign_keys.
+
+  ## Example
+
+      def article_factory do
+        %MyApp.Article{title: "An Awesome Article", author: build(:author)}
+      end
+
+      # Inserts an author and returns
+      # %{title: "An Awesome Article", author_id: 12}
```
Comment (paulcsmith) [inline lib/ex_machina/ecto.ex]: This could probably fit on one line and might read better that way

File: lib/ex_machina/ecto.ex [hunk 11]
```diff
@@ -66,6 +73,50 @@ defmodule ExMachina.Ecto do
...
+            association.owner_key,
+            association_name,
+            record
+          )
+        _ -> record
+      end
+    end)
+  end
+
+  def insert_built_belongs_to_assoc(module, foreign_key, association_name, record) do
```
Comment (paulcsmith) [inline lib/ex_machina/ecto.ex]: maybe instead of `module` it could be `repo`?

Comment (paulcsmith) [inline lib/ex_machina/ecto.ex]: And maybe that should also be changed for the other functions. Might be better for a separate PR though. I'd be fine with this as-is

File: test/ex_machina/ecto_test.exs [hunk 12]
```diff
@@ -42,4 +42,18 @@ defmodule ExMachina.EctoTest do
...
   end
+
+  test "params_with_assocs/2 inserts belongs_to associations" do
+    assert TestFactory.params_with_assocs(:article) == %{
+      title: "My Awesome Article",
+      author_id: ExMachina.TestRepo.first!(User).id,
+    }
+  end
+
+  test "params_with_assocs/2 inserts associated fields" do
```
Comment (paulcsmith) [inline test/ex_machina/ecto_test.exs]: I'm not sure I understand this test. What is this covering that the previous one isn't?

File: test/support/models/article.ex [hunk 13]
```diff
@@ -3,6 +3,8 @@ defmodule ExMachina.Article do
 
   schema "articles" do
     field :title, :string
+
     belongs_to :author, ExMachina.User
+    belongs_to :editor, ExMachina.User
```
Comment (paulcsmith) [inline test/support/models/article.ex]: I think a comment explaining why this is here might be good. Usually I don't like comments, but without that I think it is likely someone will delete this and wonder why all the tests still pass. What do you think?

**Comment by paulcsmith** [comment]
A few comments, but this looks really good :) Good to merge whenever you're ready

File: lib/ex_machina/ecto.ex (hunk 11 above)
Comment (jsteiner) [inline lib/ex_machina/ecto.ex]: Agreed that this would be better, but should probably be part of another PR.

File: test/support/models/article.ex (hunk 13 above)
Comment (jsteiner) [inline test/support/models/article.ex]: Thoughts on this, instead? https://github.com/thoughtbot/ex_machina/pull/124/commits/9a0cc9dbeafe4d7a19d080d01dccf60ad2003572

File: test/ex_machina/ecto_test.exs [hunk 14]
```diff
@@ -43,14 +43,18 @@ defmodule ExMachina.EctoTest do
     end
   end
 
-  test "params_with_assocs/2 inserts belongs_to associations" do
+  test "params_with_assocs/2 inserts belongs_tos that are set by the factory" do
+    assert Enum.member?(ExMachina.Article.__schema__(:associations), :editor)
```
Comment (paulcsmith) [inline test/ex_machina/ecto_test.exs]: I like this, but I think extracting to a private function would be good. Something like `assert has_association_in_schema?(ExMachina.Article, :editor)`. What do you think?

File: test/ex_machina/ecto_test.exs [hunk 15]
```diff
@@ -42,4 +42,22 @@ defmodule ExMachina.EctoTest do
       TestFactory.params_for(:user_map)
     end
   end
+
+  test "params_with_assocs/2 inserts belongs_tos that are set by the factory" do
+    assert Enum.member?(ExMachina.Article.__schema__(:associations), :editor)
```
Comment (paulcsmith) [inline test/ex_machina/ecto_test.exs]: Reposting because the comment was buried for some reason:

> I like this, but I think extracting to a private function would be good. Something like assert has_association_in_schema?(ExMachina.Article, :editor). What do you think?

Also, I'm not exactly sure what "editor" looks like when it's not set. So it might be nice to have an assertion for that. That would also ensure that someone doesn't accidentally add it to the factory and remove test coverage

Comment (jsteiner) [inline test/ex_machina/ecto_test.exs]: What do you mean by `I'm not exactly sure what "editor" looks like when it's not set`? If somebody adds `:editor` to the factory this test will fail because it will be included in the hash returned by `params_with_assocs`.

Comment (paulcsmith) [inline test/ex_machina/ecto_test.exs]: Sorry I wasn't that clear about that. I meant that this test relies on `editor` being something, but it's not clear in this test what it is. You'd have to look at the schema to realize that it's defined, but not being set in the factory. So there is no way to look at this test and realize that `editor` is important in any way. So I think that you could assert that it's on the schema and give it an explanatory name like `defined_on_schema?(Article, :editor)` or something

Comment (paulcsmith) [inline test/ex_machina/ecto_test.exs]: So I guess what I'm saying is that editor is a magic value. It's not clear that it's being tested
